lib/ecto/query/builder/distinct.ex

```diff
@@ -30,11 +30,21 @@ defmodule Ecto.Query.Builder.Distinct do
   Called at runtime to verify distinct.
   """
   def distinct!(query, distinct, file, line) when is_boolean(distinct) do
-    apply(query, %Ecto.Query.QueryExpr{expr: distinct, params: [], line: line, file: file})
+    apply(query, %Ecto.Query.ByExpr{expr: distinct, params: [], line: line, file: file})
   end
   def distinct!(query, distinct, file, line) do
-    {expr, params} = Builder.OrderBy.order_by_or_distinct!(:distinct, query, distinct, [])
-    expr = %Ecto.Query.QueryExpr{expr: expr, params: Enum.reverse(params), line: line, file: file}
+    {expr, params, subqueries, aliases} =
+      Builder.OrderBy.order_by_or_distinct!(:distinct, query, distinct, [])
```

Comment:
Can we actually have subqueries inside Distinct?

Reply:
Yep! I tested it out.

Reply:
```sql
SELECT DISTINCT ON (EXISTS (SELECT * FROM accounts AS a2 WHERE a2.id = a1.id)) * FROM accounts AS a1
```

Reply:
(not that you'd want to run that actual query, but confirmed that you can make subqueries in the distinct statement, at least in postgres)